tools/python_tests/permissions_test.py

```diff
@@ -66,54 +81,88 @@ def contains_correct_files(self, my_path):
+        print(on_disk)
+        apks = filter(APK_RE.match, on_disk)
+        if len(apks) < 1:
+            raise IOError("Couldn't find an APK")
+        apk_path = path.join(apk_path, apks[0])
         logging.info("Using apk at {}".format(apk_path))
 
         return apk_path
 
 
+    def get_property(self, prop, props_path):
+        if not isfile(props_path):
+            raise IOError("Properties file does not exist: {}".format(props_path))
+
+        logging.info("local props: {}".format(props_path))
+        with open(props_path) as props:
+            for line in filter(PROP_RE.match, props):
+                if line.startswith(prop):
+                    return line.split("=")[1].strip()
+
+        raise IOError("Couldn't find property {} in file {}".format(prop, props_path))
+
+
     def find_aapt(self):
-        local_props_path = join(self.omim_path, "android", "local.properties")
-        logging.info("local props: {}".format(local_props_path))
-        with open(local_props_path) as props:
-            for line in filter(lambda s: s != "" and not s.startswith("#"), map(lambda s: s.strip(), props)):
-                if line.startswith("sdk.dir"):
-                    path = line.split("=")[1].strip()
-                    aapt_path = join(path, "platforms", "android-6", "tools", "aapt")
-                    logging.info("Using aapt at {}".format(aapt_path))
-                    return aapt_path
-        logging.info("Couldn't find aapt")
-        return None
+        local_props_path = path.join(self.omim_path, "android", "local.properties")
+        sdk_path = self.get_property("sdk.dir", local_props_path)
+        aapt_path = self.find_aapt_in_platforms(sdk_path)
+
+        logging.info("Using aapt at {}".format(aapt_path))
+        return aapt_path
+
+
+    def find_aapt_in_platforms(self, platforms_path):
+        aapts = {}
+        candidates = exec_shell("find", "{} -name aapt".format(platforms_path))
+        for c in candidates:
+            if "build-tools/{}".format(self.build_tools_version) in c:
+                return c
+
+            try:
+                version = tuple(map(int, self.get_aapt_version(c)))
+                aapts[version] = c
+            except ValueError:
+                # Do nothing, because aapt version contains non-numeric symbols
+                pass
 
+        max_version = sorted(aapts.iterkeys(), reverse=True)[0]
+        logging.info("Max aapt version: {}".format(max_version))
+
+        return aapts[max_version]
+
+
+    def get_aapt_version(self, candidate):
+        return SPLIT_RE.split(exec_shell(candidate, "v")[0][31:])
 
 
     def test_permissions(self):
         """Check whether we have added or removed any permissions"""
         permissions = exec_shell(self.aapt, "dump permissions {0}".format(self.apk))
+        permissions = filter(
+            lambda s: s.startswith("permission:") or s.startswith("uses-permission:"),
+            permissions
```

Comment:
I can also inline the following "map", but it will turn into an excessively long one-liner which will be less readable.